api/application/src/review/create.rs

```diff
@@ -1,79 +1,64 @@
 use diesel::prelude::*;
-use domain::models::review::{NewReview, NewReviewWithId, Review};
+use domain::models::review::{NewReview, Review};
 use infrastructure::ServerState;
 use rocket::{http::Status, response::status::Created, serde::json::Json, State};
 use shared::{response_models::ResponseMessage, token::JWT};
 
+use crate::auth::has_user_permissions;
+use crate::database_helpers::unit_exists;
+
 pub fn create_review(
     review: Json<NewReview>,
     state: &State<ServerState>,
     token: JWT,
 ) -> Result<Created<String>, (Status, Json<ResponseMessage>)> {
     use domain::schema::reviews;
-    use domain::schema::units;
-
-    let pooled = &mut state.db_pool.get().unwrap();
 
     let review = review.into_inner();
-    let review = NewReviewWithId::new(token.claims.sub, review);
 
-    // Check if unit being review exists
-    let unit_count: i64 = match pooled.transaction(|c| {
-        units::table
-            .select(units::all_columns)
-            .filter(units::unit_code.eq(&review.unit_code))
-            .count()
-            .load(c)
-    }) {
-        Ok(unit_count) => unit_count[0],
-        Err(err) => match err {
-            diesel::result::Error::NotFound => {
+    if let Err(err) = has_user_permissions(&token, &review.user_id) {
+        return Err(err);
+    }
+
+    let pooled = &mut state.db_pool.get().unwrap();
+
+    match unit_exists(&review.unit_code, pooled) {
+        Ok(exists) => {
+            if !exists {
                 let response = ResponseMessage {
                     message: Some(String::from("Unit does not exist.")),
                 };
 
                 return Err((Status::NotFound, Json(response)));
             }
-            _ => {
-                panic!("Database error - {}", err);
-            }
-        },
+        }
+        Err(err) => return Err(err),
```

Comment:
Since we're just early returning the error and taking the inner value we can use `?`, e.g. `if !unit_exists(&review.unit_code, pooled)? {`